server/src/main/java/org/apache/druid/metadata/SQLMetadataSegmentManager.java

```diff
@@ -219,82 +219,112 @@ public void stop()
     }
   }
 
+  private VersionedIntervalTimeline<String, DataSegment> getVersionedIntervalTimeline(final String dataSource)
```

Comment:
Is it necessary to construct the entire timeline? I think you could restrict the segments in the timeline here to just those that overlap interval-wise with the target interval or the intervals of the target segments

Reply:
@jon-wei This was something I was wondering. Also should it be restricted to segments that are enabled or will be enabled? For instance if a segment is overshadowed by a disabled segment should it still be enabled?

Reply:
I think you could consider only the currently enabled segments that overlap the target intervals, since only that set would potentially overshadow things.

> For instance if a segment is overshadowed by a disabled segment should it still be enabled?

If the timeline is built from only currently enabled segments, then I think this case would only arise when a user calls the segmentId-based API, and the list they submit contains segments that would overshadow each other. For that, I think this sequence would work:
- Build the timeline from currently enabled segments
- Add all the target segments to this timeline
- For each target segment, call isOvershadowed and only enable the ones where that returns false

Reply:
The timeline should now only be created with segments required to determine overshadowed-ness.